README.md

```diff
@@ -104,6 +104,166 @@ TODO: specify the encoding (byte-array to string) procedure
 
 TODO: specify the decoding (string to byte-array) procedure
 
+### Codecs
+
+Depending on the protocol of a Multiaddr item, different algorithms are used to
+convert their values from/to binary representation. The name of the codec to use
+for each protocol is noted in [protocols.csv](protocols.csv).
+
+In general empty values in the string representation are always disallowed unless
+explicitely noted otherwise. In case of conversion errors implementation must
+refuse to process the given string/binary value and report the error to the caller
+instead.
+
+While all codec types may be encoded using the standard variable length encoding,
+some may also allow being encoded into a specific static-length binary value
+without the extra length information if this is noted in the respective codec's
+description.
+
+All code examples are written in Python-based pseudo code and are optimized for
+legibility rather than speed. In general you should always use existing libraries
+and functions for performing the below conversions rather than rolling your own.
+
+#### `fspath`
+
+Encodes the given Unicode string using the system's local file system encoding.
+On Windows this encoding likely being UTF-16, while being UTF-8 on most other
+systems. It is up to the library to figure out the best encoding value for
+these kinds of strings.
+
+ * String → Binary: `str.encode(SYSTEM_FILESYSTEM_ENCODING)`
+ * Binary → String: `bytes.decode(SYSTEM_FILESYSTEM_ENCODING)`
+
+Protocols using the `fspath` encoding must not be shared between different hosts.
+
+#### `idna`
```

Comment:
https://github.com/multiformats/multiaddr/issues/94

Reply:
Think about this again, what I'm actually trying to achieve with this is proper canonicalization of the input. And while IDNA encoding has all that there is no reason why can't use UTS-46 normalization with UTF-8 encoding. So how about this instead:

> Encodes the given Unicode representation to the UTF-8 character encoding ([RFC 3629 Section 3](https://tools.ietf.org/html/rfc3629#section-3)), while using the [UTS-46 / RFC 5890](https://tools.ietf.org/html/rfc5890) input normalization and processing rules for canonicalization.
> 
> * String → Binary:
>    1. If feasible, normalize and validate the given input string according to [UTS-46 Section 4 (Processing)](https://www.unicode.org/reports/tr46/#Processing) and [UTS-46 Section 4.1 (Validity Criteria)](https://www.unicode.org/reports/tr46/#Validity_Criteria) with the following parameters:
>       * UseSTD3ASCIIRules = true
>       * CheckHyphens = true
>       * CheckBidi = true
>       * CheckJoiners = true
>       * Transitional_Processing = false
>    2. Convert the Unicode string to the UTF-8 character encoding as per [RFC 3629 Section 3 §4](https://tools.ietf.org/html/rfc3629#section-3).
> * Binary → String:  
>   Convert the UTF-8 encoded binary string to Unicode according to the rules of [RFC 3629 Section 3 §6](https://tools.ietf.org/html/rfc3629#page-5).

Reply:
That sounds like a great idea (and it fixes the bytewise comparison problem).